File: finger-detection/photo_puzzle.py
"""
FingerQuest Photo Puzzle Module
================================
Game 3: user draws a rectangular area on the live camera feed with their
index finger, that area is captured as an image, auto-shuffled into a 3x3
grid, and the user rearranges the tiles back into place.

Interaction:
  1. Hold your index finger still over the first corner (~0.6s) to lock it.
  2. Move to the opposite corner and hold still again to lock it and capture.
  3. Drag tiles (same grab/drag/snap mechanic as the numbered puzzle) back
     into the 3x3 target grid.
"""

import cv2
from enum import Enum, auto

from puzzle import create_photo_puzzle, PuzzleBoard

WHITE = (255, 255, 255)
GREEN = (0, 255, 136)
YELLOW = (0, 255, 255)
CYAN = (255, 255, 0)
DARK_BG = (20, 20, 30)


class PhotoPuzzleState(Enum):
    CORNER1 = auto()
    CORNER2 = auto()
    PLAYING = auto()


class AreaSelector:
    """Locks two corners of a rectangle via finger dwell (hold-still)."""

    DWELL_FRAMES = 18       # ~0.6s at 30fps
    DWELL_RADIUS = 18       # px jitter tolerance
    MIN_SIZE = 160           # minimum captured region (px) for decent tiles

    def __init__(self):
        self.corner1 = None
        self.corner2 = None
        self._dwell_pos = None
        self._dwell_frames = 0

    @property
    def done(self) -> bool:
        return self.corner1 is not None and self.corner2 is not None

    def dwell_progress(self) -> float:
        return min(self._dwell_frames / self.DWELL_FRAMES, 1.0)

    def update(self, finger_x: int, finger_y: int, finger_visible: bool):
        if self.done or not finger_visible:
            self._dwell_pos = None
            self._dwell_frames = 0
            return

        if self._dwell_pos is None:
            self._dwell_pos = (finger_x, finger_y)
            self._dwell_frames = 1
        else:
            dx = finger_x - self._dwell_pos[0]
            dy = finger_y - self._dwell_pos[1]
            if dx * dx + dy * dy <= self.DWELL_RADIUS ** 2:
                self._dwell_frames += 1
            else:
                self._dwell_pos = (finger_x, finger_y)
                self._dwell_frames = 1

        if self._dwell_frames >= self.DWELL_FRAMES:
            self._lock(finger_x, finger_y)
            self._dwell_pos = None
            self._dwell_frames = 0

    def _lock(self, x: int, y: int):
        if self.corner1 is None:
            self.corner1 = (x, y)
        else:
            x1, y1 = self.corner1
            if abs(x - x1) < self.MIN_SIZE:
                x = x1 + self.MIN_SIZE if x >= x1 else x1 - self.MIN_SIZE
            if abs(y - y1) < self.MIN_SIZE:
                y = y1 + self.MIN_SIZE if y >= y1 else y1 - self.MIN_SIZE
            self.corner2 = (x, y)

    def get_rect(self, frame_w: int, frame_h: int):
        """Return (x1, y1, x2, y2) clamped to frame bounds, or None."""
        if not self.done:
            return None
        x1, y1 = self.corner1
        x2, y2 = self.corner2
        x1, x2 = sorted((x1, x2))
        y1, y2 = sorted((y1, y2))
        x1, y1 = max(x1, 0), max(y1, 0)
        x2, y2 = min(x2, frame_w), min(y2, frame_h)
        return (x1, y1, x2, y2)

    def draw(self, img, finger_x: int, finger_y: int, finger_visible: bool):
        h, w = img.shape[:2]

        overlay = img.copy()
        cv2.rectangle(overlay, (0, 0), (w, h), (10, 10, 20), -1)
        cv2.addWeighted(overlay, 0.25, img, 0.75, 0, img)

        if self.corner1 is not None:
            cv2.circle(img, self.corner1, 8, GREEN, -1)
            cv2.circle(img, self.corner1, 8, WHITE, 2)

        if self.corner1 is not None and finger_visible:
            cv2.rectangle(img, self.corner1, (finger_x, finger_y), CYAN, 2)

        if finger_visible:
            cv2.circle(img, (finger_x, finger_y), 10, YELLOW, 2)
            if self._dwell_pos is not None:
                progress = self.dwell_progress()
                cv2.ellipse(img, (finger_x, finger_y), (18, 18), -90, 0,
                            360 * progress, GREEN, 3)

        instruction = ("Tahan diam di sudut PERTAMA area"
                        if self.corner1 is None else
                        "Tahan diam di sudut KEDUA area")
        (tw, _), _ = cv2.getTextSize(instruction, cv2.FONT_HERSHEY_SIMPLEX,
                                      0.6, 2)
        cv2.putText(img, instruction, (w // 2 - tw // 2, 40),
                    cv2.FONT_HERSHEY_SIMPLEX, 0.6, WHITE, 2, cv2.LINE_AA)


class PhotoPuzzleGame:
    """Draw an area with your finger, then solve a 3x3 puzzle of it."""

    def __init__(self, frame_w: int, frame_h: int):
        self.frame_w = frame_w
        self.frame_h = frame_h
        self.state = PhotoPuzzleState.CORNER1
        self.selector = AreaSelector()
        self.board: PuzzleBoard | None = None
        self.solution_thumb = None
        self._last_finger = (-1, -1, False)

    def update(self, finger_x: int, finger_y: int, finger_visible: bool, frame):
        self._last_finger = (finger_x, finger_y, finger_visible)
        if self.state != PhotoPuzzleState.PLAYING:
            self.selector.update(finger_x, finger_y, finger_visible)
            if self.selector.corner1 is not None and self.state == PhotoPuzzleState.CORNER1:
                self.state = PhotoPuzzleState.CORNER2
            if self.selector.done:
                self._capture(frame)
        else:
            self.board.update(finger_x, finger_y, finger_visible)

    def _capture(self, frame):
        x1, y1, x2, y2 = self.selector.get_rect(self.frame_w, self.frame_h)
        crop = frame[y1:y2, x1:x2].copy()
        self.board = create_photo_puzzle(self.frame_w, self.frame_h, crop)
        self.solution_thumb = cv2.resize(crop, (90, 90))
        self.state = PhotoPuzzleState.PLAYING

    def reshuffle(self):
        if self.board:
            self.board.shuffle()

    def draw(self, img):
        if self.state != PhotoPuzzleState.PLAYING:
            fx, fy, finger_visible = self._last_finger
            self.selector.draw(img, fx, fy, finger_visible)
            return

        self.board.draw(img)

        # Solution thumbnail preview (top-left)
        if self.solution_thumb is not None:
            h, w = self.solution_thumb.shape[:2]
            x, y = 20, 55
            cv2.rectangle(img, (x - 3, y - 3), (x + w + 3, y + h + 3),
                          WHITE, 2)
            img[y:y + h, x:x + w] = self.solution_thumb
            cv2.putText(img, "Target", (x, y - 10),
                        cv2.FONT_HERSHEY_SIMPLEX, 0.45, WHITE, 1, cv2.LINE_AA)

    def draw_status(self, img):
        if self.state == PhotoPuzzleState.PLAYING and self.board:
            self.board.draw_status(img)

File: web/src/game/photoPuzzle.ts
import { AREA_DWELL_MS, COLOR } from "./constants";
import { createPhotoPuzzle, PuzzleBoard } from "./puzzle";
import type { FingerState } from "./types";

const DWELL_RADIUS = 18;
const MIN_SIZE = 160;

type Corner = { x: number; y: number } | null;

class AreaSelector {
  corner1: Corner = null;
  corner2: Corner = null;
  private dwellPos: { x: number; y: number } | null = null;
  private dwellMs = 0;

  get done(): boolean {
    return this.corner1 !== null && this.corner2 !== null;
  }

  dwellProgress(): number {
    return Math.min(this.dwellMs / AREA_DWELL_MS, 1);
  }

  tick(finger: FingerState, dtMs: number) {
    if (this.done || !finger.visible) {
      this.dwellPos = null;
      this.dwellMs = 0;
      return;
    }

    if (!this.dwellPos) {
      this.dwellPos = { x: finger.x, y: finger.y };
      this.dwellMs = dtMs;
    } else {
      const dx = finger.x - this.dwellPos.x;
      const dy = finger.y - this.dwellPos.y;
      if (dx * dx + dy * dy <= DWELL_RADIUS * DWELL_RADIUS) {
        this.dwellMs += dtMs;
      } else {
        this.dwellPos = { x: finger.x, y: finger.y };
        this.dwellMs = dtMs;
      }
    }

    if (this.dwellMs >= AREA_DWELL_MS) {
      this.lock(finger.x, finger.y);
      this.dwellPos = null;
      this.dwellMs = 0;
    }
  }

  private lock(x: number, y: number) {
    if (!this.corner1) {
      this.corner1 = { x, y };
    } else {
      let lx = x;
      let ly = y;
      if (Math.abs(x - this.corner1.x) < MIN_SIZE) {
        lx = x >= this.corner1.x ? this.corner1.x + MIN_SIZE : this.corner1.x - MIN_SIZE;
      }
      if (Math.abs(y - this.corner1.y) < MIN_SIZE) {
        ly = y >= this.corner1.y ? this.corner1.y + MIN_SIZE : this.corner1.y - MIN_SIZE;
      }
      this.corner2 = { x: lx, y: ly };
    }
  }

  getRect(frameW: number, frameH: number) {
    if (!this.done || !this.corner1 || !this.corner2) return null;
    const x1 = Math.max(0, Math.min(this.corner1.x, this.corner2.x));
    const y1 = Math.max(0, Math.min(this.corner1.y, this.corner2.y));
    const x2 = Math.min(frameW, Math.max(this.corner1.x, this.corner2.x));
    const y2 = Math.min(frameH, Math.max(this.corner1.y, this.corner2.y));
    return { x: x1, y: y1, w: x2 - x1, h: y2 - y1 };
  }

  draw(
    ctx: CanvasRenderingContext2D,
    w: number,
    h: number,
    finger: FingerState
  ) {
    ctx.fillStyle = "rgba(10, 10, 20, 0.25)";
    ctx.fillRect(0, 0, w, h);

    if (this.corner1) {
      ctx.fillStyle = COLOR.green;
      ctx.beginPath();
      ctx.arc(this.corner1.x, this.corner1.y, 8, 0, Math.PI * 2);
      ctx.fill();
      ctx.strokeStyle = COLOR.white;
      ctx.lineWidth = 2;
      ctx.stroke();
    }

    if (this.corner1 && finger.visible) {
      ctx.strokeStyle = COLOR.cyan;
      ctx.lineWidth = 2;
      ctx.strokeRect(
        this.corner1.x,
        this.corner1.y,
        finger.x - this.corner1.x,
        finger.y - this.corner1.y
      );
    }

    if (finger.visible) {
      ctx.strokeStyle = COLOR.yellow;
      ctx.lineWidth = 2;
      ctx.beginPath();
      ctx.arc(finger.x, finger.y, 10, 0, Math.PI * 2);
      ctx.stroke();

      if (this.dwellPos) {
        const progress = this.dwellProgress();
        ctx.strokeStyle = COLOR.green;
        ctx.lineWidth = 3;
        ctx.beginPath();
        ctx.arc(finger.x, finger.y, 18, -Math.PI / 2, -Math.PI / 2 + Math.PI * 2 * progress);
        ctx.stroke();
      }
    }

    const instruction = this.corner1 === null
      ? "Tahan diam di sudut PERTAMA area"
      : "Tahan diam di sudut KEDUA area";
    ctx.fillStyle = COLOR.white;
    ctx.font = "bold 18px sans-serif";
    ctx.textAlign = "center";
    ctx.fillText(instruction, w / 2, 40);
    ctx.textAlign = "left";
  }
}

export class PhotoPuzzleGame {
  private frameW: number;
  private frameH: number;
  state: "CORNER1" | "CORNER2" | "PLAYING" = "CORNER1";
  private selector = new AreaSelector();
  board: PuzzleBoard | null = null;
  private solutionThumb: HTMLCanvasElement | null = null;
  private lastFinger: FingerState = { x: -1, y: -1, visible: false };

  constructor(frameW: number, frameH: number) {
    this.frameW = frameW;
    this.frameH = frameH;
  }

  update(finger: FingerState, dtMs: number, cleanCanvas: HTMLCanvasElement) {
    this.lastFinger = finger;
    if (this.state !== "PLAYING") {
      this.selector.tick(finger, dtMs);
      if (this.selector.corner1 !== null && this.state === "CORNER1") {
        this.state = "CORNER2";
      }
      if (this.selector.done) {
        this.capture(cleanCanvas);
      }
    } else {
      this.board?.update(finger, dtMs);
    }
  }

  private capture(cleanCanvas: HTMLCanvasElement) {
    const rect = this.selector.getRect(this.frameW, this.frameH);
    if (!rect) return;
    this.board = createPhotoPuzzle(
      this.frameW,
      this.frameH,
      cleanCanvas,
      rect.x,
      rect.y,
      rect.w,
      rect.h
    );

    const thumb = document.createElement("canvas");
    thumb.width = 90;
    thumb.height = 90;
    thumb
      .getContext("2d")!
      .drawImage(cleanCanvas, rect.x, rect.y, rect.w, rect.h, 0, 0, 90, 90);
    this.solutionThumb = thumb;

    this.state = "PLAYING";
  }

  reshuffle() {
    this.board?.shuffle();
  }

  draw(ctx: CanvasRenderingContext2D) {
    if (this.state !== "PLAYING") {
      this.selector.draw(ctx, this.frameW, this.frameH, this.lastFinger);
      return;
    }

    this.board?.draw(ctx);

    if (this.solutionThumb) {
      const x = 20;
      const y = 55;
      ctx.strokeStyle = COLOR.white;
      ctx.lineWidth = 2;
      ctx.strokeRect(x - 3, y - 3, 96, 96);
      ctx.drawImage(this.solutionThumb, x, y, 90, 90);
      ctx.fillStyle = COLOR.white;
      ctx.font = "12px sans-serif";
      ctx.fillText("Target", x, y - 10);
    }
  }

  drawStatus(ctx: CanvasRenderingContext2D) {
    if (this.state === "PLAYING" && this.board) {
      this.board.drawStatus(ctx, this.frameW, this.frameH);
    }
  }
}
